internal/runtime/registry/registry.go

```diff
@@ -1,55 +1,224 @@
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	return extensions.ready
 }
 
-func (r registry) RegisterRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Add all the Runtime Extensions served by a registered runtime extension server.
+// Please note that if the provided registration object already exists, all the registered
+// RuntimeExtensions gets updated according to the newly provide status.
+func (extensions *extensionRegistry) Add(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("invalid argument, when calling Add ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+	return extensions.add(ext)
 }
 
-func (r registry) RemoveRuntimeExtension(ext *runtimev1.Extension) {
-	panic("implement me")
+// Remove all the Runtime Extensions served by a registered runtime extension server.
+func (extensions *extensionRegistry) Remove(ext *runtimev1.Extension) error {
+	if !extensions.ready {
+		return errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	if ext == nil {
+		return errors.New("invalid argument, when calling Remove ext must not be nil")
+	}
+
+	extensions.lock.Lock()
+	defer extensions.lock.Unlock()
+
+	extensions.remove(ext)
+	return nil
 }
 
-func (r registry) GetRuntimeExtension(gvh catalog.GroupVersionHook, name string) RuntimeExtensionRegistration {
-	panic("implement me")
+func (extensions *extensionRegistry) remove(ext *runtimev1.Extension) {
+	for _, e := range extensions.items {
+		if e.RegistrationName == ext.Name {
+			delete(extensions.items, e.Name)
+		}
+	}
 }
 
-func (r registry) GetRuntimeExtensions(gvh catalog.GroupVersionHook) []RuntimeExtensionRegistration {
-	panic("implement me")
+// List all the registered Runtime Extensions for a given Group/Hook..
+func (extensions *extensionRegistry) List(gh catalog.GroupVersionHook) ([]*RuntimeExtensionRegistration, error) {
+	if !extensions.ready {
+		return nil, errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	l := make([]*RuntimeExtensionRegistration, 0, 10)
+	for _, r := range extensions.items {
+		if r.GroupVersionHook.Group == gh.Group && r.GroupVersionHook.Hook == gh.Hook {
+			l = append(l, r)
+		}
+	}
+	return l, nil
 }
 
-type Registry interface {
-	RemoveRuntimeExtension(ext *runtimev1.Extension)
-	RegisterRuntimeExtension(ext *runtimev1.Extension)
+// Get the Runtime Extensions with a given name.
+func (extensions *extensionRegistry) Get(name string) (*RuntimeExtensionRegistration, error) {
+	if !extensions.ready {
+		return nil, errors.New("invalid operation: Get cannot called on a registry not yet ready")
+	}
+
+	extensions.lock.RLock()
+	defer extensions.lock.RUnlock()
+
+	r, _ := extensions.items[name]
```

Comment:
Correct me if I am wrong, isn't the combination of `r.RegistrationName` and `r.Name` supposed to be unique and just `r.Name` is not guaranteed to be unique across all the runtime extensions?

Same comment on the `add` function in l.220.

Reply:
So the full story is like this:
* An Extension with name "my-amazing-extensions" is created in the kube-apiserver
* The extension controller will discover the extension
* The Discovery response will have unique names, e.g. "http-proxy"
* The extension controller will add the discovered RuntimeExtensions with e.g. `http-proxy.my-amazing-extension` to the status

So r.Name is for example `http-proxy.my-amazing-extension` which is also the format we pass into the `Get` func.

So we only have the plan `http-proxy` name in the DiscoveryReponse.

Reply:
Discussed with Yuvaraj, should be fine as is